src/AggregateFunctions/AggregateFunctionFunnelAnalysis.cpp

```diff
@@ -0,0 +1,58 @@
+#include <AggregateFunctions/AggregateFunctionFactory.h>
+#include <AggregateFunctions/AggregateFunctionFunnelAnalysis.h>
+#include <AggregateFunctions/Helpers.h>
+#include <AggregateFunctions/FactoryHelpers.h>
+#include <DataTypes/DataTypeDate.h>
+#include <DataTypes/DataTypeDateTime.h>
+#include <DataTypes/DataTypeArray.h>
+
+#include <ext/range.h>
+#include "registerAggregateFunctions.h"
+
+
+namespace DB
+{
+namespace ErrorCodes
+{
+    extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+    extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+}
+
+namespace
+{
+
+template <template <typename> class Data>
+AggregateFunctionPtr createAggregateFunctionFunnelAnalysis(const std::string & name, const DataTypes & arguments, const Array & params)
+{
+    if (params.empty())
+        throw Exception{"Aggregate function " + name + " requires at least three parameters: <window>, <strict>, <strict_order>, [option, [option, ...]]", ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH};
+
+    if (arguments.size() < 2)
+        throw Exception("Aggregate function " + name + " requires one event argument and one timestamps argument.", ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH);
+
+    if (params.size() > max_events + 3)
+        throw Exception("Too many event parameters for aggregate function " + name, ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH);
+
+    DataTypePtr ts_data_type_ptr = assert_cast<const DataTypeArray *>(arguments[1].get())->getNestedType();
```

Comment:
It would be better to use `typeid_cast<const DataTypeArray *>` and check that cast is successful (not nullptr is returned), if not - throw an exception with an appropriate message.

Reply:
make sense, will update.